include/nuttx/semaphore.h

```diff
@@ -59,6 +59,48 @@
      {(c), (f), SEM_WAITLIST_INITIALIZER}
 #endif /* CONFIG_PRIORITY_INHERITANCE */
 
+/* Fast mutex lock/unlock path */
+
+#if defined(CONFIG_PRIORITY_PROTECT) || defined(CONFIG_PRIORITY_INHERITANCE)
+#define NXSEM_IF_FASTMUTEX(sem)                                   \
+  (((sem)->flags & (SEM_TYPE_MUTEX | SEM_PRIO_MASK)) ==           \
+   (SEM_TYPE_MUTEX | SEM_PRIO_NONE))
+#else
+#define NXSEM_IF_FASTMUTEX(sem) ((sem)->flags & SEM_TYPE_MUTEX)
+#endif
+
+/* Try wait; sets ret to OK or -EAGAIN when success,
+ * -EPERM when fast locking can't be performed
+ */
+
+#define NXSEM_FAST_TRYWAIT_IF_MUTEX(sem, ret)                           \
+  {                                                                     \
+    ret = -EPERM;                                                       \
+    if (NXSEM_IF_FASTMUTEX(sem))                                        \
+      {                                                                 \
+        int old = 1;                                                    \
+        ret = atomic_try_cmpxchg_acquire(NXSEM_COUNT(sem), &old, 0) ?   \
+          OK : -EAGAIN;                                                 \
+      }                                                                 \
+  }
+
+/* Post; sets ret to OK when success and -EPERM when
+ * fast posting can't be performed
+ */
+
+#define NXSEM_FAST_POST_IF_MUTEX(sem, ret)                              \
```

Comment:
One option would be to use a statement expression, but I'm not sure if using those is legal in common code (like semaphores), as those are GCC only..